proposals/module-linking/Binary.md

```diff
@@ -0,0 +1,238 @@
+
+A new module section is added
+
+```
+aliassec ::=  a*:section_102(vec(alias))     ->        a*
+
+alias ::=
+    0x00 i:instanceidx 0x00 e:exportidx      ->       (alias (instance $i) (func $e))
+    0x00 i:instanceidx 0x01 e:exportidx      ->       (alias (instance $i) (table $e))
+    0x00 i:instanceidx 0x02 e:exportidx      ->       (alias (instance $i) (memory $e))
+    0x00 i:instanceidx 0x03 e:exportidx      ->       (alias (instance $i) (global $e))
+    0x00 i:instanceidx 0x05 e:exportidx      ->       (alias (instance $i) (module $e))
+    0x00 i:instanceidx 0x06 e:exportidx      ->       (alias (instance $i) (instance $e))
+    0x01 0x05 m:moduleidx                    ->       (alias parent (module $m))
+    0x01 0x07 t:typeidx                      ->       (alias parent (type $t))
+```
+
+**Validation**
+
+* Aliased instance indexes are all in bounds
+* Aliased instance export indices are in bounds relative to the instance's
+  *locally-declared* (via module or instance type) list of exports
+* Export indices match the actual type of the export
+* Aliases append to the respective index space.
+* Parent aliases can only happen in submodules (not the top-level module) and
+  the index specifies is the entry, in the parent's raw index space, for that
+  item.
+* Parent aliases can only refer to preceeding module/type definitions, relative
+  to the binary location where the inline module's type was declared. Note that
+  when the module code section is defined all of the parent's modules/types are
+  available, but inline modules still may not have access to all of these if the
+  items were declared after the module's type in the corresponding module
+  section.
+
+## Function section
+
+**Validation**
+
+* Type indices must point to function types.
+
+## Export section
+
+update
+[`exportdesc`](https://webassembly.github.io/spec/core/binary/modules.html#binary-exportdesc)
+
+```
+exportdesc ::=
+    ...
+    0x05 x:moduleidx                        -> module x
+    0x06 x:instanceidx                      -> instance x
+```
+
+**Validation**
+
+* Module/instance indexes must be in-bounds.
+
+## Module Code Section (103)
+
+```
+modulecodesec ::= m*:section_103(vec(modulecode))       ->    m*
+
+modulecode ::= size:u32 mod:module                      ->    mod, size = ||mod||
+```
+
+Note that this is intentionally a recursive production where `module` refers to
+the top-level
+[`module`](https://webassembly.github.io/spec/core/binary/modules.html#binary-module)
+
+**Validation**
+
+* Module definitions must match their module type exactly, no subtyping (or
+  maybe subtyping, see WebAssembly/module-linking#9).
+* Modules themselves validate recursively.
+* Must have the same number of modules as the count of all local module
+  sections.
+* Each submodule is validated with a subset of the parent's context, for example
+  the set of types and instances the current module has defined are available
+  for aliasing in the submodule.
```

Comment:
As far as the spec is concerned, validation will probably not be concerned with the module dec / module code section separation (like it isn't for functions). So in the specification, it will just validate the inner modules with the current incremental outer context.